types/connector.d.ts

```diff
@@ -3,28 +3,285 @@
+  fkName: string | null,
+  fkTableName: string,
+  fkColumnName: string,
+  keySeq: number,
+
+  pkOwner: string | null,
+  pkName: string | null,
+  pkTableName: string,
+  pkColumnName: string,
+}
+
+export type DiscoveredModelProperties = {
+  owner?: string,
+  tableName?: string,
+  columnName?: string,
+  dataType?: string,
+  dataLength?: number,
+  dataPrecision?: number,
+  dataScale?: number,
+  nullable?: boolean,
+}
+
+
+// #TODO(achrinza): The codebase suggets that `context` differs
+// depending on the situation, and that there's no cohesive interface.
+// Hence, we'll need to identify all the possible contexts.
+export type Context = {
+  Model: ModelBaseClass,
+  instance?: object,
+  query?: Filter,
+  where?: Where,
+  data?: AnyObject,
+  hookState?: AnyObject,
+  options?: Options,
+  isNewInstance?: boolean,
+  currentInstance?: Readonly<ModelBase>,
+}
 
 /**
  * Connector from `loopback-connector` module
  */
 export interface Connector {
   name: string; // Name/type of the connector
-  dataSource?: DataSource;
-  connect(callback?: Callback): PromiseOrVoid; // Connect to the underlying system
-  disconnect(callback?: Callback): PromiseOrVoid; // Disconnect from the underlying system
-  ping(callback?: Callback): PromiseOrVoid; // Ping the underlying system
+
+  /**
+   * @internal
+   */
+  _models?: Record<string, ModelBaseClass>;
+  connect?(callback?: Callback): PromiseOrVoid; // Connect to the underlying system
+  disconnect?(callback?: Callback): PromiseOrVoid; // Disconnect from the underlying system
+
+  /**
+   * Ping the underlying connector to test the connections.
+   * 
+   * @remarks
+   * Unlike {@link DataSource.ping}, if no callback is provided, a
+   * {@link Promise} return value is not guaranteed.
+   * 
+   * @param callback Callback function
+   * @returns a {@link Promise} or `void`
+   */
+  ping?(callback?: Callback): PromiseOrVoid; // Ping the underlying system
   execute?(...args: any[]): Promise<any>;
+
+  /**
+   * Get the connector's types collection.
+   * 
+   * @remarks
+   * For example, ['db', 'nosql', 'mongodb'] would be represent a datasource of
+   * type 'db', with a subtype of 'nosql', and would use the 'mongodb' connector.
+   *
+   * Alternatively, ['rest'] would be a different type altogether, and would have
+   * no subtype.
+   * 
+   * @returns The connector's type collection.
+   */
+  getTypes?(): string[];
+  define?(def: {model: ModelBaseClass, properties: PropertyDefinition, settings: ModelDefinition['settings']}): void;
+
+  /**
+   * Define a property on the target model.
+   * 
+   * @param model Name of model
+   * @param prop Name of property
+   * @param params Property settings
+   */
+  defineProperty?(model: string, prop: string, params: PropertyDefinition): void;
+  defineForeignKey?(modelName: string, key: string, foreignModelName: string, cb: Callback<PropertyType>): void;
+  defineForeignKey?(modelName: string, key: string, cb: Callback<PropertyType>): void;
```

Comment:
There's several of these functions that accept a callback, but I'm not sure if its optional.